dashboard/app/controllers/registrations_controller.rb

```diff
@@ -181,4 +244,40 @@ def update_params(params)
       races: []
     )
   end
+
+  def set_email_params
+    if params[:user][:email_preference_opt_in].present?
+      params[:user][:email_preference_request_ip] = request.env['REMOTE_ADDR']
+      params[:user][:email_preference_source] = EmailPreference::ACCOUNT_EMAIL_CHANGE
+      params[:user][:email_preference_form_kind] = "0"
+    end
+    params.require(:user).
+      permit(
+        :email,
+        :hashed_email,
+        :current_password,
+        :email_preference_opt_in,
+        :email_preference_request_ip,
+        :email_preference_source,
+        :email_preference_form_kind,
+      )
+  end
+
+  def set_user_type_params
+    if params[:user][:email_preference_opt_in].present?
+      params[:user][:email_preference_request_ip] = request.env['REMOTE_ADDR']
+      params[:user][:email_preference_source] = EmailPreference::ACCOUNT_TYPE_CHANGE
+      params[:user][:email_preference_form_kind] = "0"
+    end
```

Comment:
nit: would it make sense to DRY up these repeated lines (249-253 and 267-271) into another method?
Maybe copy `params` instead of mutating it, just for good measure?

Also if the base `user` param is missing, this line 267 will raise NoMethodError, causing a 500 internal server error, whereas `require(:user)` will raise `ActionController::ParameterMissing` causing a 400 Bad Request. 

It could be something like this: 
```ruby
def user_params_with_email
  params.require(:user).tap do |user_params|
    user_params[:email_preference_request_ip] = request.env['REMOTE_ADDR']
    # ...
  end
end

def set_email_params
  user_params_with_email.permit(
    :email,
    # ...
  )
end
```

Reply:
I'll give this a shot.

Reply:
Very cool!  I puzzled over how to DRY the `permit()` call for a bit, bit I think I've got it.  The trick was to `merge()` in the cleaned up email preference parameters!

```rb
  def set_email_params
    params.
      require(:user).
      permit(
        :email,
        :hashed_email,
        :current_password,
      ).
      merge(email_preference_params(EmailPreference::ACCOUNT_EMAIL_CHANGE, "0"))
  end

  def set_user_type_params
    params.
      require(:user).
      permit(
        :user_type,
        :email,
        :hashed_email,
      ).
      merge(email_preference_params(EmailPreference::ACCOUNT_TYPE_CHANGE, "0"))
  end

  def email_preference_params(source, form_kind)
    params.
      require(:user).
      tap do |user|
        if user[:email_preference_opt_in].present?
          user[:email_preference_request_ip] = request.env['REMOTE_ADDR']
          user[:email_preference_source] = source
          user[:email_preference_form_kind] = form_kind
        end
      end.
      permit(
        :email_preference_opt_in,
        :email_preference_request_ip,
        :email_preference_source,
        :email_preference_form_kind,
      )
  end
```

Reply:
Nice!